src/pages/billing/_tests/CreateAzureBillingProjectModal.test.js

```diff
@@ -0,0 +1,236 @@
+  })
+
+  const tooShortError = 'Billing project name is too short (minimum is 6 characters)'
+  const tooLongError = 'Billing project name is too long (maximum is 30 characters)'
+  const badCharsError = 'Billing project name can only contain letters, numbers, underscores and hyphens.'
+  const duplicateProjectError = 'Billing project name already exists'
+  const getBillingProjectInput = () => screen.getByLabelText('Terra billing project *')
+  const getBillingProjectHint = () => screen.queryByText('Name must be unique and cannot be changed.')
+
+  const invalidUuidError = 'Subscription id must be a UUID'
+  const noManagedApps = 'No Terra Managed Applications exist for that subscription'
+  const managedAppCallFailed = 'Unable to retrieve Managed Applications for that subscription'
+  const getSubscriptionInput = () => screen.getByLabelText('Azure subscription *')
+  const getManagedAppInput = () => screen.getByLabelText('Managed application *')
+  const getCreateButton = () => screen.getByText('Create')
+
+  const verifyDisabled = item => expect(item).toHaveAttribute('disabled')
+  const verifyEnabled = item => expect(item).not.toHaveAttribute('disabled')
+
+  it('has the correct initial state', () => {
+    // Assert
+    expect(getBillingProjectHint()).not.toBeNull()
+    verifyDisabled(getManagedAppInput())
+    verifyDisabled(getCreateButton())
+  })
+
+  it.each([
+    { text: 'short', errors: [tooShortError] },
+    { text: '', errors: [tooShortError, badCharsError] },
+    { text: '**********', errors: [badCharsError] },
+    { text: 'aowejfawioefjaowiejfaoiwejfoaijw efoiawjefiajwefijaweoifjaweoijf', errors: [badCharsError, tooLongError] },
+    { text: 'ThisIs-a_suitableName', errors: [] }
+  ])('validates billing project name "$text"', ({ text, errors }) => {
+    // Arrange
+    const allErrors = [tooShortError, tooLongError, badCharsError, duplicateProjectError]
+    // Act
+    if (text === '') {
+      // Must first insert something and then replace with empty string.
+      fireEvent.change(getBillingProjectInput(), { target: { value: 'temp' } })
+    }
+    fireEvent.change(getBillingProjectInput(), { target: { value: text } })
+    // Assert
+    if (_.isEmpty(errors)) {
+      expect(getBillingProjectHint()).not.toBeNull()
+      _.forEach(unexpectedError => expect(screen.queryByText(unexpectedError)).toBeNull(), allErrors)
+    } else {
+      expect(getBillingProjectHint()).toBeNull()
+      _.forEach(error => expect(screen.queryByText(error)).not.toBeNull(), errors)
+      _.forEach(unexpectedError => expect(screen.queryByText(unexpectedError)).toBeNull(), _.difference(allErrors, errors))
+    }
+    // Create always disabled because all fields must have valid inputs.
+    verifyDisabled(getCreateButton())
+  })
+
+  it('validates the subscription ID', () => {
+    // Arrange
+    expect(screen.queryByText(invalidUuidError)).toBeNull() // Should not initially be visible
```

Comment:
to move this assertion down to the // Assert section I recommend you save the result of screen.queryByText(...) to a const initialXxxxError here and assert it .toBeNull() in the later section.

Reply:
@msilva-broad to be honest, I don't really care too much for the AAA pattern. As the tests become more complicated, you are "asserting" at multiple points and "acting" at multiple points. "Arranging" and even "acting" may also happen in test setup. I think it is easier to understand the test if you can see, "this is what I expect the state to be at this point of the test", vs. storing variables and doing all the assertions at the end.

Reply:
a compromise I have seen work well is to have additional // Assert and // Act sections as needed for these types of scenarios.

// Assert
expect(...)

// Act
...

// Assert
expect(...)

A goal of the AAA pattern is to have reasonably crisp, small tests that give good signal on what broke if they fail.  I hope to introduce additional design patterns that should make it easier to avoid large multi-step test scenarios since the code under test will be more modularized and testable as smaller units.   Having to do AAAAAA in large tests is usually a code-smell that the code under test could use some refactoring to be easier to test.

But I am not one to let perfect be the enemy of good.  AAAA or whatever is a good compromise since it keeps things well labeled and readable for possible refactoring later.

Reply:
@msilva-broad this component is new and small-- it's just 3 fields. However, the fields depend on each other for state, and the interaction between them is what I am most interested in testing. This results in multi-step test scenarios, but I still think they can be very readable without requiring AAA labels. I would rather see comments or self-documenting code that describe the setup (which is what I think of as `Arrange`, where setup includes validating the expected initial state), the actions that are being triggered, and the results being validated.

I am going to push up a commit where I remove the AAA labels and instead increase the comments. I can revert this or go to a mixture of AAA labels and comments, but this will let you see more concretely how I would write the tests if I wasn't trying to follow the pattern of including AAA labels.

Reply:
to summarize our video chat on this topic:
- renaming this section of the test as:
     // Assert - Initial State
   should work.
- the AAA or AAAAA labels are very helpful for onboarding developers who are less experienced in writing good unit/functional tests, and also helpful for folks reviewing the tests, which will have some variations (helper methods, beforeEach, it.each, etc) in large test suites, and help reviewers who are less familiar with the code area in general.

Reply:
@msilva-broad I pushed up the changes we talked about.